site/docs/expressions/scalar_functions.md

```diff
@@ -43,10 +43,68 @@ For scalar functions (one value produced for each record), a function binding de
 
 
 
+## Argument Types & Output Type Derivation
+
+### Input Argument Types
+
+Input arguments can be declared in one of three ways: complete, wildcarded, parameterized.
+
+* Complete Type: Complete types are either simple types or compound types with all parameters fully defined. Examples include i32, fp32, VARCHAR(20), List&lt;fp32&gt;, etc.
+* Wildcard Types: Compound types can be wildcarded meaning that any valid value for the parameter can be accepted. Types are wildcarded by failing to express a particular value for the type. When serialized, this will typically mean using a null pointer or sentinel value. Examples would be  VARCHAR(*), List&lt;\*&gt;, DECIMAL(\*,10).
+* Parameterized types: Types can be parameterized such that the parameter can be used in a output derivation. Example parameterized types would be VARCHAR(T), List&lt;E&gt;, MAP&lt;K, fp32&gt;, etc. Parameters are simple names. Any parameters defined in an argument must be referenced at least once in an output expression. Parameterized argument types can only be used when a function is using expression based output type derivation.
```

Comment:
> Any parameters defined in an argument must be referenced at least once in an output expression.

What is the rationale here?

Reply:
Per my comment above. A wildcard could also be called an unnamed parameter. A parameter is something that must be used. It is only to be as intentional as possible.

Reply:
Hm, I guess I see this as analogous to an unused variable. It seems like introducing a new concept only to prevent people from not using type variables is pretty heavy weight since now someone has to build support for both, as well as the enforcement mechanism, when you can unify the notions of wildard and parameterized types under a single concept.

Reply:
Part of it was thinking about the serialization formats, which I guess we can ignore. When serializing something it seems weird to require a named unused parameter. If you say it is required at the spec level (e.g. you don't allow List<>, only List<K>) it seems like it would be confusing if the serialization then forced this on you. I suppose serialization could say either representation is fine and just leave it to implementors to have to tease out the distinction. In general, I'm trying to be more formal about these things as much as possible (and why I work on the protobuf sketch as I think about the problems).

I'm okay switching it but that was the reasoning :)

Reply:
I agree with cpcloud here.  What about the case where it is unused in the output derivation, can be any type, but is not valid as a wildcard.  For example:

`EQUAL(FIXEDBINARY<*>, FIXEDBINARY<*>) -> boolean` is not valid because fixed binaries are only comparable if they have the same width.  However, the type parameters are not used in the output derivation.

Reply:
Good point on the equal. I've updated to remove the distinction between wildcard and parameterized.